File: solutions/0297. Serialize and Deserialize Binary Tree/0297-2.cpp
class Codec {
 public:
  // Encodes a tree to a single string.
  string serialize(TreeNode* root) {
    string s;
    preorder(root, s);
    return s;
  }

  // Decodes your encoded data to tree.
  TreeNode* deserialize(string data) {
    istringstream iss(data);
    queue<string> q;

    for (string s; iss >> s;)
      q.push(s);

    return preorder(q);
  }

 private:
  void preorder(TreeNode* root, string& s) {
    if (!root) {
      s += "n ";
      return;
    }

    s += to_string(root->val) + " ";
    preorder(root->left, s);
    preorder(root->right, s);
  }

  TreeNode* preorder(queue<string>& q) {
    const string s = q.front();
    q.pop();
    if (s == "n")
      return nullptr;

    TreeNode* root = new TreeNode(stoi(s));
    root->left = preorder(q);
    root->right = preorder(q);
    return root;
  }
};

File: solutions/0297. Serialize and Deserialize Binary Tree/0297-2.java
public class Codec {
  // Encodes a tree to a single string.
  public String serialize(TreeNode root) {
    StringBuilder sb = new StringBuilder();
    preorder(root, sb);
    return sb.toString();
  }

  // Decodes your encoded data to tree.
  public TreeNode deserialize(String data) {
    final String[] vals = data.split(" ");
    Queue<String> q = new ArrayDeque<>(Arrays.asList(vals));
    return preorder(q);
  }

  private void preorder(TreeNode root, StringBuilder sb) {
    if (root == null) {
      sb.append("n ");
      return;
    }

    sb.append(root.val).append(" ");
    preorder(root.left, sb);
    preorder(root.right, sb);
  }

  private TreeNode preorder(Queue<String> q) {
    final String s = q.poll();
    if (s.equals("n"))
      return null;

    TreeNode root = new TreeNode(Integer.parseInt(s));
    root.left = preorder(q);
    root.right = preorder(q);
    return root;
  }
}
